### PR TITLE
Abstract out notion of clock from timer-related methods

### PR DESCRIPTION
**Pull request by Vilkina**
- [x]  Introduce Clock class with methods for getting the current time, nano time, and sleeping
- [x]  Create production implementation (Live).
  Create test implementation (Test).
  Make retry / repeat methods and Schedule methods take a Clock w/default to Live.
  Add tests as appropriate
  Document any changes in interface

### CODE REVIEW

File: core/shared/src/main/scala/scalaz/zio/Clock.scala [hunk 1]
```diff
@@ -16,7 +16,7 @@ trait Clock {
 object Clock {
 
   object Live extends Clock {
-    def currentTime(unit: TimeUnit): IO[Nothing, Long] = IO.sync(unit.convert(System.nanoTime(), NANOSECONDS))
+    def currentTime(unit: TimeUnit): IO[Nothing, Long] = IO.sync(unit.convert(System.currentTimeMillis(), NANOSECONDS))
```
Comment (rssh) [inline core/shared/src/main/scala/scalaz/zio/Clock.scala]: hmm,  currentTimeMillis return milliseconds, but in second argument of convert is NANOSECONDS.
Is this a typo ?

**Comment by Vilkina** [comment]
Yes, you are right this a typo.

06 сентября 2018, 11:33:42, от "Ruslan Shevchenko" <notifications@github.com>:

@rssh commented on this pull request.

In core/shared/src/main/scala/scalaz/zio/Clock.scala:

> @@ -16,7 +16,7 @@ trait Clock {
 object Clock {
 
   object Live extends Clock {
-    def currentTime(unit: TimeUnit): IO[Nothing, Long] = IO.sync(unit.convert(System.nanoTime(), NANOSECONDS))
+    def currentTime(unit: TimeUnit): IO[Nothing, Long] = IO.sync(unit.convert(System.currentTimeMillis(), NANOSECONDS))
hmm, currentTimeMillis return milliseconds, but in second argument of convert is NANOSECONDS.
Is this a typo ?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.



--
Best regards,

Oleksandra Holubitska

**Comment by jdegoes** [comment]
@Vilkina Looks great, thanks for your first pull request! 🙏 🎉

**Comment by Vilkina** [comment]
@jdegoes will finish this issue soon.

**Pull request merged**